### PR TITLE
DO NOT MERGE - Approve with nullable annotations

### PR DESCRIPTION
**Pull request by blairconrad**
For discussion. Maybe we can understand what's going on here.

### CODE REVIEW

File: tests/FakeItEasy.Tests.Approval/ApiApproval.ApproveApiNetStd20.approved.txt [hunk 1]
```diff
@@ -7,62 +7,69 @@
...
 {
+    [System.Runtime.CompilerServices.NullableAttribute(0)]
     public class static A
     {
         public static FakeItEasy.Configuration.IVoidArgumentValidationConfiguration CallTo(System.Linq.Expressions.Expression<System.Action> callSpecification) { }
         public static FakeItEasy.Configuration.IAnyCallConfigurationWithNoReturnTypeSpecified CallTo(object fake) { }
-        public static FakeItEasy.Configuration.IReturnValueArgumentValidationConfiguration<T> CallTo<T>(System.Linq.Expressions.Expression<System.Func<T>> callSpecification) { }
-        public static FakeItEasy.Configuration.IPropertySetterAnyValueConfiguration<TValue> CallToSet<TValue>(System.Linq.Expressions.Expression<System.Func<TValue>> propertySpecification) { }
-        public static System.Collections.Generic.IList<T> CollectionOfDummy<T>(int numberOfDummies) { }
+        public static FakeItEasy.Configuration.IReturnValueArgumentValidationConfiguration<T> CallTo<[System.Runtime.CompilerServices.NullableAttribute(2)]  T>(System.Linq.Expressions.Expression<System.Func<T>> callSpecification) { }
```
Comment (blairconrad) [inline tests/FakeItEasy.Tests.Approval/ApiApproval.ApproveApiNetStd20.approved.txt]: the typeparam is marked as "annotated" for some reason, which I guess means... nullable? If I attempt to call the method with a `Expression<Func<string?>>` I get no warning.
Confirmed! If I constrain by `notnull`, I get a warning in my call with an `Expression<Func<string?>>`.

File: tests/FakeItEasy.Tests.Approval/ApiApproval.ApproveApiNetStd20.approved.txt [hunk 2]
```diff
@@ -7,62 +7,69 @@
...
+        public static T Contains<[System.Runtime.CompilerServices.NullableAttribute(0)]  T>(this FakeItEasy.IArgumentConstraintManager<T> manager, object value)
             where T : System.Collections.IEnumerable { }
         public static string EndsWith(this FakeItEasy.IArgumentConstraintManager<string> manager, string value) { }
         public static System.Threading.CancellationToken IsCanceled(this FakeItEasy.IArgumentConstraintManager<System.Threading.CancellationToken> manager) { }
-        public static T IsEmpty<T>(this FakeItEasy.IArgumentConstraintManager<T> manager)
+        public static T IsEmpty<[System.Runtime.CompilerServices.NullableAttribute(0)]  T>(this FakeItEasy.IArgumentConstraintManager<T> manager)
             where T : System.Collections.IEnumerable { }
-        public static T IsEqualTo<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, T value) { }
-        public static T IsGreaterThan<T>(this FakeItEasy.IArgumentConstraintManager<T> manager, T value)
+        public static T IsEqualTo<[System.Runtime.CompilerServices.NullableAttribute(2)]  T>(this FakeItEasy.IArgumentConstraintManager<T> manager, T value) { }
```
Comment (blairconrad) [inline tests/FakeItEasy.Tests.Approval/ApiApproval.ApproveApiNetStd20.approved.txt]: I understand why this `T` is nullable here, but not why it's "oblivious" in `IsEmpty`…

**Comment by blairconrad** [comment]
For fun, requiring VS 2019 Preview to get the 3.0.0 SDK preview …

**Comment by thomaslevesque** [comment]
> Build succeeded.
>    0 Warning(s)
 >   0 Error(s)

Isn't that strange? With nullable ref types enabled, there should have plenty of warnings. But on the other hand, it did emit type annotations...
In earlier previews of the SDK, the property to set to enable it was `NullableContextOptions` instead of `Nullable`. I think AppVeyor doesn't have the latest SDK, because in Hamlet I [had to set the old property](https://github.com/thomaslevesque/Hamlet/commit/61470035c415270c999c5e6ce50786a6c1d8b025).

**Comment by thomaslevesque** [comment]
> But on the other hand, it did emit type annotations...

Ah, but these annotations were generated on your machine, not on AppVeyor...

**Comment by thomaslevesque** [comment]
> I think AppVeyor doesn't have the latest SDK

Confirmed. As mentioned in chat, the VS2019 Preview image has .NET Core SDK 3.0.100-preview-009812, which is the very first preview of .NET Core 3 SDK.

**Comment by blairconrad** [comment]
Well, I enjoyed talking about this. Time to close the review, I think.

**Pull request closed without merging**